### DISCUSSION TITLE
Share your ideas! What JointJS or JointJS+ demos would you like us to create?

### DISCUSSION THREAD

**Discussion by kumilingus** [clientIO/joint#1643] [Ideas]
We post new demos every Wednesday, and while we’re full of ideas, your contribution would mean the world to us. Do you have a specific use case in mind or would you like to see some data expressed visually? Share your ideas and we’ll be happy to include them in our Demo Wednesday initiative, just as we’ve done with dozens of app demos that come from our customers and community members.

It’s simple, just follow these three steps:
1. Describe the use case/idea in the comment.
2. Provide the source of the data (if relevant).
3. Share a diagram that could serve as inspiration (optional).

And if you're looking for inspiration, check out the top 15 community-designed demo apps: [https://www.jointjs.com/blog/15-community-designed-demo-apps](https://www.jointjs.com/blog/15-community-designed-demo-apps).

We look forward to seeing your ideas! 😊

**Comment by bl-akonwie** [comment]
I love this puzzle demo
 
https://resources.jointjs.com/demos/puzzle

At this point, I'd like a more interesting puzzle instead of a common jigsaw.
The demo is currently splitting the original image into polygon pieces. 
Instead of that, can we split the image into more of a blobby/wavy circle where all of them fit together to create a whole of the original image?

Thanks!

**Reply by kumilingus** [reply]
Thank you for reaching out! Is this an example of what you are looking for?
<img width="305" alt="image" src="https://user-images.githubusercontent.com/3967880/169256650-e5d6eee1-e618-4d6f-8e11-08d8e540b117.png">

**Comment by eddhib** [comment]
is it possible to design a page or dashboard with widgets like Jira (reorder, resize, nested container, grid 12 columns layout) with Jointjs, if so ideally see a demo

**Reply by kumilingus** [reply]
Could please paste an image or a link?

**Reply by eddhib** [reply]
Hello,
there is a tool that does that https://gridstackjs.com/#demo, 
will jointjs be able to respond to the same feature?

**Reply by kumilingus** [reply]
JointJS does not have any masonry layouts available. If there is an open source layout library that accepts the order of elements and their sizes and returns the positions of the elements, it would be possible to create a similar demo.

Please let me ask you a few questions:
- What would be the benefit of rendering the tiles in SVG as opposed to HTML containers with an SVG widgets inside?
- Why don't you use the `gridstackjs` library and e.g. put `JointJS` papers inside the tiles?

**Reply by pfalcon64** [reply]
JointJs' ability to save an entire drawing to SVG is invaluable for a dashboard component, as it means a user can save the current view directly to a Slide. This means you cannot use a combination of HTML and SVG - it MUST be pure SVG

**Reply by kumilingus** [reply]
You can use export to `PNG` or `Canvas`.

**Comment by igo** [comment]
I would like to see existing demo https://resources.jointjs.com/demos/sequence with more interactive features like how to add a new connection between actors.

**Reply by kumilingus** [reply]
Implemented as part of [HoverConnect Tool Pull Request](https://github.com/clientIO/joint/pull/1763). Available in a couple of months in `v3.6` (or right now on `master` branch).

![sequence](https://user-images.githubusercontent.com/3967880/183423920-e4515fb2-004e-485e-a5d2-368e784166c0.gif)

**Reply by kumilingus** [reply]
It's available [here](https://resources.jointjs.com/demos/sequence) now.

**Comment by vidhanbagdai** [comment]
My organization uses web to create BPMN 2.0 shapes and has started exploring the BPMN Visio export app provided by JointJS+ library. Although in the app provided in the trial version, we could see export for only basic shapes such as start, end, activity, sequence flow, etc. We would love to see a demo of export to some complicated shapes as well like Subprocess, pools and lanes, text annotations, etc.

**Comment by AlexMath23** [comment]
Hi, 
We recognise that you can do a lot of customisation with the build in jointjs shapes / svg's. 
But for our company's use case those were not enough for the dynamic elements that we want to show. 
We now have React components that we render on the paper, and this works but it was/is a hassle to get it working properly. 
=> using portals to place components into foreignObject svg's in our custom cells and making sure that the graph/paper and the components stay synced, + a bunch off little quirks we had to take care off. What would really improve our experience / serve our needs is first class component (react) support from the library it self. Maybe something like a jointjs-react fork could be possible in the future ( 4.0 and beyond) when jquery / lodash get cycled out ?. It would also greatly enhance customizability/look & feel if you could use your own components with the existing logic for selectbox,menu,tooltip,popup,... etc. 

Love the library and would be awesome if this could be included. 

Best regards,
Alex

**Reply by kumilingus** [reply]
Hi @AlexMath23, sorry for the late reply. I'm afraid this is not the correct channel for your request. This thread is for ideas for new demos. What you are asking for is a new feature/module/library.

Feel free to create a new issue or feature request. Also, if you have anything on how to use the React portals with JointJS, it would be great if you could share it there as well. Thank you.

**Reply by kumilingus** [reply]
Here is an early version of the `@joint/react` library: https://github.com/clientIO/joint/pull/2391
It's not a fork, JointJS is a dependency. Any and all feedback would be appreciated.

**Reply by AlexAtlas** [reply]
Very cool Roman !! , I'll dive a bit deeper in this PR if I find some time this week. I'm also looking forward to the 4.0 release !

**Reply by kumilingus** [reply]
Thanks @AlexAtlas , insight from someone experienced in React would be awesome!

**Reply by igo** [reply]
Looking at 
```js
    const paper = usePaper();
    const zoomIn = () => {
        if (!paper) return;
        const { sx, sy } = paper.scale();
        paper.scale(sx * 2, sy * 2);
    }
```

Using  `if (!paper) return;` is not a good practice. You can also create more sophisticated hooks. Take a look for example at https://react-hook-form.com/docs/useform. You can create hooks like this:

```js
import { usePaper, Paper } from 'joinjs/react';

const { paper, zoomIn, zoomOut, ...manyOtherFunctionsOrObjects } = usePaper(graph);

<button onClick={zoomIn}>Zoom in</button>
<Paper paper={paper} />
```

**Reply by kumilingus** [reply]
1) What is the good practice? Throw an exception?
2) It's not my intention to add functions here that are essentially just syntactic sugar for actual paper methods. It's meant to be a low-level component, so I don't see a use for it here (yet). But it is definitely something for higher-level components such as <Diagram /> and `useDiagram()` hook (as mentioned in the README).

*If you don't mind, please add your next comments directly to the PR*

**Comment by dkchaitanya92** [comment]
Hi @kumilingus,

I want to have a communication between two different elements, where one element is present in container A and another element is present in container B. 

Explanation (Each element will have different shape, Each element will have "in" and "out" ports as well, container will not have any ports):
 
1. Element1, Element2, Element3, Link1(Element1 to Element 2), Link2(Element 2 to Element 3) will be present in a container A.
2. Element4, Element5, Element6, Link3(Element 4 to Element 5), Link4(Element 5 to Element 6) will be present in a container B.
3. There should be a connection between container A and container B. Connection Link from Element 3 to Element 4.
4. I want to show text on the connection link which is shown between two containers as well.
5. If I minimize the container then want to load the image.

Reference - https://www.jointjs.com/demos/container.


Kindly provide the example for Typescript and React using JointJS.
(Graphical representation to be achieved dynamically with JSON input).

Thanks in advance.

**Reply by kumilingus** [reply]
Hi @dkchaitanya92, thank you for reaching out to us!

Could you please elaborate on no. 5?

Just to confirm, when the container is collapsed, the links between containers should reconnect from the container children to the parent i.e minimized container? Similar to our [Activity Diagram](https://codepen.io/jointjs/pen/PodyMbm)? Is that correct?

Also I am curious what kind of diagram is this? Does it have a name? What data does it show? Thank you.

**Reply by dkchaitanya92** [reply]
Hi @kumilingus 

When container A is collapsed and container B isn't collapsed then we need to show all the child Links connected to container A. 

Please find Image below.

![NodeCollapsed](https://user-images.githubusercontent.com/47445361/232017825-e41e3385-d296-4886-b6ed-ad8fb9ec3362.PNG)

If both container A and container B is in collapsed form then two scenario's to be handled.
Scenario 1: show only one link between the container A and container B are collapsed.
Scenario 2: When we perform double click on the link, it can expand all the links between Container A and Container B.

Diagram name is "Network Graphical View". 

Data it is displaying is networking nodes, cards and links information.

(Nodes are like container, elements are the cards and internal links are the communication between cards and external links are connection between the nodes).

Similar to our [Activity Diagram](https://codepen.io/jointjs/pen/PodyMbm) - Yes (a little bit). but here when I minimize, I will show the above image as Node Collapsed.

Kindly provide the code in Typescript and react using JointJS

**Reply by kumilingus** [reply]
Ok @dkchaitanya92 , thank you. This could be a nice demo. Please bear with us.

**Reply by dkchaitanya92** [reply]
Forgot to mention about the ports @kumilingus.

1. Each element will have "IN" ports and "OUT" ports which is present inside the container.
2. Sometimes each element can have one or two or three "IN" ports and sometimes each element can have one or two or three "OUT" ports.
3. There will be event listeners as well on each container, on each element at "IN" (OR) "OUT" ports and on each link as well.

Explanation: Why event listeners are required ?

Based on socket (Web socket / short polling / Server side events) notifications, 
1. I can update the colour of the port present in the element.
2. I can update update the colour of Link present between two ports.
3. I can update the alarm present on the Node as well.

So, event listeners will help me to update all of the above three points.

**Reply by kumilingus** [reply]
Okay @dkchaitanya92, we can spice up the demo by animating it a little bit.
What do the colors indicate? Is there any source of information about this type of diagram that you could share with us? I haven't had any luck Googling it (it's important for naming the variables and make it understandable for anyone else looking at the example).

**Reply by dkchaitanya92** [reply]
Hi @kumilingus 

**Information source** : 
1. https://www.youtube.com/watch?v=xvs1rtz2too
2. https://community.fs.com/blog/an-overview-of-dwdm-technology-and-dwdm-system-components.html
3. https://en.wikipedia.org/wiki/Wavelength-division_multiplexing

I am getting the data from the Nodes in real time, So, my team wants to use the data:
1. To show the circuit in the graphical representation to operator, which is convenient to track or perform the operations on the circuit.
2. Graphical representation will help the operator during maintenance when switching from Work path to protect path (OR) vice versa.

In real time, the naming conventions are same - Nodes, Cards, Ports, Links and FIBER Links(connection between two containers).

**Reply by kumilingus** [reply]
Wow, that's very helpful! We'll start working on it soon. Please note that this is only a demo, so we won't try to implement all the things you've described here. But the demo should definitely get you started.

**Reply by kumilingus** [reply]
Here's the [DWDM](https://github.com/clientIO/joint/discussions/2163) demo. It's written in Open Source JointJS and Typescript.

1. _JJ+_ is a superset of _JJ_. That means that you can just replace _JJ_ with _JJ+_ and everything will work.
2. To integrate the example with React please follow our [tutorial](https://resources.jointjs.com/tutorial/react-ts).
3. To add a tree index on the left side please see the [Diagram Index](https://www.jointjs.com/demos/diagram-index) example (React version).
4. I can imagine adding the _JJ+_ plugin [ui.Navigator](https://resources.jointjs.com/docs/rappid/v3.6/ui.Navigator.html), to allow users to quickly navigate through large circuits.

**Reply by MeAriJit** [reply]
@kumilingus Great demo! Helpful. Just wanted to know where did you get the Node assets svg images? Did you create on your own?

**Reply by kumilingus** [reply]
Thank you. They were free icons with a permissive license downloaded from an icons marketplace.

**Reply by MeAriJit** [reply]
Could you share me the details of the marketplace?

**Reply by kumilingus** [reply]
https://www.iconfinder.com/

**Comment by saravntbe** [comment]
Demo for rendering cells > 200 in canvas with better performance. Right now we tried setting `async` option,  but it doesn't solve performance issue.

**Reply by kumilingus** [reply]
Here's a demo with 2000 cells: https://www.jointjs.com/demos/collapse-expand

**Comment by Habib2959** [comment]
Hello @kumilingus,
Can I create an organization chart like the image with joint js or joint js+? if yes, then may I get some insights?
![IMG_20230828_162607](https://github.com/clientIO/joint/assets/71340861/ca0545d2-76d9-477c-aab4-a8604ce7daac)

**Reply by kumilingus** [reply]
Yes, you can. Use the [layout.TreeLayout](https://resources.jointjs.com/docs/rappid/v3.7/layout.TreeLayout.html) to layout only the elements forming the tree structure at the top (similarly what we did [here](https://codepen.io/jointjs/pen/LYrQvKZ?editors=0010)). Then you can layout the rest of the children (with a layout of your choice) and position elements them under the middle of related branch.

We're happy to create an example of this if you are able to:
- provide us with an example data
- explain how such a diagram is used

**Reply by Habib2959** [reply]
Thank you for nthe response. I will provide a demo of this. This would be a pleasure if you create an example with that.

**Reply by Habib2959** [reply]
@kumilingus The chart will be similar to the image. If Joint js can serve our purpose, we will go for Joint JS+. Thank you.

![chart](https://github.com/clientIO/joint/assets/71340861/0c13623f-ce28-4268-879d-b471849b5571)

**Reply by kumilingus** [reply]
Before we can write an example, we need to understand how the diagram is used and we need the data to be displayed. The image you shared unfortunately just adds more confusion.

**Comment by gonzalo-dominguez** [comment]
@kumilingus I'm not sure if this is the correct place to ask, does JointJS support bowtie visualization out of the box? Is there any available example or guidance you could provide on this topic?

Thanks in advance 🙇

**Reply by kumilingus** [reply]
I imagine that you can create custom elements and use the `TreeLayout` to automatically position them.
<img width="419" alt="image" src="https://github.com/user-attachments/assets/7ed7b5b9-841e-46ab-99fe-15c909964482" />

Do you have any specific **bowtie** implementation in mind? Could you share an image?
